File: src/helpers/position.rs
#[derive(Clone)]
pub struct Position {
	pub index: usize,
	pub line: usize,
	col: usize,
	pub filename: String
}

impl Position {
	pub fn new(filename: String, index: usize, line: usize, col: usize) -> Self {
		Position { filename, index, line, col }
	}

	pub fn advance(&mut self, cc: char) {
		self.index += 1;
		self.col += 1;

		if cc == '\n' {
			self.line += 1;
			self.col = 0;
		}
	}
}

File: src/helpers/position.cpp
#include "../headers/position.hpp"

Position::Position(const size_t index, const size_t line, const size_t col, const std::string& filename, const std::string& text) : index(index), line(line), col(col), filename(filename), text(text) {}

void Position::advance(const char cc) {
	index++;
	col++;

	// Next line
	if(cc == '\n') {
		line++;
		index = 0; // Reset index on changing line
		col = 0;
	}
}

